nextcord/http.py

```diff
@@ -155,25 +164,283 @@ def bucket(self) -> str:
+        exc_tb: Optional[TracebackType],
     ) -> None:
-        if self._unlock:
-            self.lock.release()
+        self.release()
+
+    def locked(self) -> bool:
+        return self.remaining <= 0
+
+    async def acquire(self) -> bool:
+        # If no more requests can be made but the event is set, clear it.
+        if self.remaining <= 0 and self._on_reset_event.is_set():
+            _log.debug(
+                "Bucket %s: Hit the remaining request limit of %s, locking until reset.",
+                self.bucket,
+                self.limit,
+            )
+            self._on_reset_event.clear()
+            if not self.resetting:
+                self.start_reset_task()
+
+        # Waits in a loop for the event to be set, clearing the event as needed and looping.
+        while not self._on_reset_event.is_set():
+            _log.debug("Bucket %s: Not set yet, waiting for it to be set.", self.bucket)
+            await self._on_reset_event.wait()
+
+            if self.remaining <= 0 and self._on_reset_event.is_set():
+                _log.debug(
+                    "Bucket %s: Hit the remaining limit of %s, locking until reset.",
+                    self.bucket,
+                    self.limit,
+                )
+                self._on_reset_event.clear()
+                if not self.resetting:
+                    self.start_reset_task()
+
+        if self.migrating:
+            raise RateLimitMigrating(
+                f"This RateLimit is deprecated, you need to migrate to bucket {self.migrating}"
+            )
+        elif self._deny:
+            raise ValueError("This request path 404'd and is now denied.")
+
+        _log.debug("Bucket %s: Continuing with request.", self.bucket)
+        self.remaining -= 1
+        return True
+
+    def release(self) -> None:
+        # Basically a placeholder, could probably be removed ;)
+        pass
+
+
+class GlobalRateLimit(RateLimit):
+    """
+    Represents the global rate limit, and thus has to have slightly modified behavior.
+
+    Still not thread safe.
+    """
+
+    async def acquire(self) -> bool:
+        ret = await super().acquire()
+        # As updates are little weird, it's best to start the reset task as soon as the first request has acquired.
+        if not self.resetting:
+            self.start_reset_task()
+
+        return ret
+
+    async def update(self, response: aiohttp.ClientResponse) -> None:
+        if response.headers.get("X-RateLimit-Global") != "true":
+            # The response is intended for the regular rate limit, not a global rate limit.
+            return
+
+        if response.status == 429:
+            # Oh dear, we hit the rate limit.
+            _log.warning("Global rate limit 429 encountered, setting remaining to 0.")
+            self.remaining = 0
+            data = await response.json()
+            _log.warning(data)
+            if response.headers.get("X-RateLimit-Scope") == "global":
```

Comment:
Not really? This specifically checks if the rate limit is in the global scope. I'm checking it because I don't trust Discord to not do some funky thing where it's a global rate limit, but it's not in the global scope for some reason?
Also adjusted where `await response.json()` is ran to be inside of the `if` statement.

Reply:
I guess that could be possible, depending on how the value of `X-RateLimit-Global` is calculated...

Reply:
Duh, fixing.